### PR TITLE
[Host] - Player Confidence Graph 

### PR DESCRIPTION
**Pull request by drewjhart**
_Note: this is PR is one of a series of nested PRs that provide new features to `host`. They breakdown as such:
[[Host] - Base UI Updates in Advance of New Features #792](https://github.com/rightoneducation/righton-app/pull/792)
->[[Host] - Real Time Responses #793](https://github.com/rightoneducation/righton-app/pull/793)
->->[[Host] - Adds API call to Confidence Switch action #799](https://github.com/rightoneducation/righton-app/pull/799)
->->->[[Host] - Player Confidence Graph #800 ](https://github.com/rightoneducation/righton-app/pull/800)
->->->->[[Host] - Prettier #804](https://github.com/rightoneducation/righton-app/pull/804)_

**Issue:**

This PR merges Alex's branch: [https://github.com/rightoneducation/righton-app/tree/alex--host-confidence-meter-UI](alex--host-confidence-meter-UI) into the ongoing transition `host` modifications. It also moves around some of the functions into `HelpFunctions.jsx` and adapts the ModuleNavigator in the footer to respond to `isConfidenceEnabled`. Finally, it cleans up some of the code and optimizes some of the sorting functions that handle how the Confidence Graph displays data.

**Preview: **

You can see a preview of the final result of the nested PRs in the below video. This includes the Real Time Responses graph and the Confidence Meter features and allows for the Confidence feature to be turned on and off on a per question basis:

https://github.com/rightoneducation/righton-app/assets/79417944/5d4985c8-252a-4170-afef-d0178d84d779

### CODE REVIEW

File: host/src/containers/GameSessionContainer.tsx [hunk 1]
```diff
@@ -187,33 +232,40 @@ const GameSessionContainer = () => {
   const handleBeginQuestion = () => {
     // I'm keeping this console.log in until we figure out NOT_STARTED so we can tell there's been a change in state 
     console.log(gameSession.currentState);
+    const gameSessionId = gameSession?.id;
+    const questionId = gameSession.questions[gameSession.currentQuestionIndex].id;
+    const order = gameSession.questions[gameSession.currentQuestionIndex].order;
```
Comment (maniramezan) [inline host/src/containers/GameSessionContainer.tsx:237]: Why one line uses `gameSession?` and not the other ones? Can also use `isNullOrDefault` and do early exist which ensures `gameSession` is not `null` and Typescript guarantees it as well

Comment (maniramezan) [inline host/src/containers/GameSessionContainer.tsx:237]: Personal opinion: Usually when accessing the same path multiple times, e.i. `gameSession.questions[gameSession.currentQuestionIndex]`, better to extract it into a variable and then use that variable:

```typescript
    const currentQuestion = gameSession.questions[gameSession.currentQuestionIndex];
    const questionId = currentQuestion.id;
    const order = currentQuestion.order;
```

File: host/src/containers/GameSessionContainer.tsx [hunk 2]
```diff
@@ -187,33 +232,40 @@ const GameSessionContainer = () => {
...
-
-          Promise.all(teamDataRequests)
-            .then(responses => {
-              setTeamsArray(responses);
-            })
-            .catch(reason => console.log(reason));
-        });
-      setIsTimerActive(true);
-      setIsLoadModalOpen(true);
+            if (gameSession.currentState === GameSessionState.TEAMS_JOINING) {
```
Comment (maniramezan) [inline host/src/containers/GameSessionContainer.tsx:241]: Maybe do an early exist here by changing it to `!==` and return. That helps the rest of the code below to be indented one less

File: host/src/containers/GameSessionContainer.tsx [hunk 3]
```diff
@@ -187,33 +232,40 @@ const GameSessionContainer = () => {
...
+                      setGameSession(response);
+                      const teamDataRequests = response.teams.map(team => {
+                        return apiClient.getTeam(team.id); // got to call the get the teams from the API so we can see the answers
+                      });
+            
+                      Promise.all(teamDataRequests)
+                        .then(responses => {
+                          setTeamsArray(responses);
+                        })
+                        .catch(reason => console.log(reason));
```
Comment (maniramezan) [inline host/src/containers/GameSessionContainer.tsx:260]: `console.error` probably

File: host/src/lib/HelperFunctions.jsx [hunk 4]
```diff
@@ -100,6 +120,15 @@ export const getAnswersByQuestion = (choices, teamsArray, currentQuestionIndex,
               choices && choices.forEach(choice => {
                 if (answer.text === choice.text) {
                   answersArray[choicesTextArray.indexOf(choice.text)] += 1;
+                  const index = confidenceLevelsArray.indexOf(answer.confidenceLevel);
+                  if (choice.isAnswer){
```
Comment (maniramezan) [inline host/src/lib/HelperFunctions.jsx:124]: nit: Space before `{`

**Review by maniramezan — APPROVED** [review]

File: networking/src/ApiClient.ts [hunk 5]
```diff
@@ -839,6 +867,17 @@ class TeamAnswerParser {
...
     }
 
+    static teamAnswerFromUpdateTeamAnswerSubscription(
+        subscription: OnUpdateTeamAnswerSubscription
+    ): ITeamAnswer {
+        const updateTeamAnswer = subscription.onUpdateTeamAnswer
+        if (isNullOrUndefined(onCreateTeamAnswer)) {
+            throw new Error("subscription.onCreateTeamAnswer can't be null.")
+        }
+        //@ts-ignore
```
[the remark this answers is not in the document]
Comment (drewjhart) [reply networking/src/ApiClient.ts:877]: You're right, we don't. I've corrected the null check in 874 to createTeamAnswer and then remove the ts-ignore.

File: host/src/components/ConfidenceResponses/ConfidenceResponseDropdown.jsx [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+  Card
+} from "@material-ui/core";
+import check from '../../images/correctAnswerCheck.png';
+
+export default function ConfidenceResponseDropdown({ 
+  graphClickInfo, 
+  selectedConfidenceData 
+}) {
+  const classes = useStyles();
+   // TODO: integrate this into ConfidenceLevel enum to prevent use of dictionaries here and in confidenceresponsegraph
```
[the remark this answers is not in the document]
Comment (drewjhart) [reply host/src/components/ConfidenceResponses/ConfidenceResponseDropdown.jsx:15]: Good idea, I've created a github issue for this and provided the link on a second line in this comment

File: host/src/components/ConfidenceResponses/ConfidenceResponseDropdown.jsx [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    );
+  };
+  // sorts players based on the Figma criteria:
+  // correct players sorted alphabetically
+  // incorrect players sorted first by answer frequency, then alphabetically
+  const sortPlayers = (selectedConfidenceData) => {
+    const correctPlayers = [];
+    const incorrectPlayers = [];
+    const answerFrequency = {};
+    selectedConfidenceData.players.map((playerData) => {
```
[the remark this answers is not in the document]
Comment (drewjhart) [reply host/src/components/ConfidenceResponses/ConfidenceResponseDropdown.jsx:42]: nice, good catch

File: host/src/components/ConfidenceResponses/ConfidenceResponseDropdown.jsx [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+      // split players into correct and incorrect so .sort is limited to these subsets
+      if (playerData.isCorrect) {
+        correctPlayers.push(playerData);
+      } else {
+        incorrectPlayers.push(playerData);
+        // if incorrect, also store the frequency of the answer for sorting later
+        answerFrequency[playerData.answer] = (answerFrequency[playerData.answer] || 0) + 1;
+      };
+    });
+    // sort correct alphabeticall
```
[the remark this answers is not in the document]
Comment (drewjhart) [reply host/src/components/ConfidenceResponses/ConfidenceResponseDropdown.jsx:52]: yup, good eyes!

File: host/src/components/ConfidenceResponses/ConfidenceResponseGraph.jsx [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+  // TODO: integrate this into ConfidenceLevel enum to prevent use of dictionaries here and in confidenceresponsedropdown
+  const ConfidenceLevelDictionary = {
+    0: "Not\nRated",
+    1: "Not At\nAll",
+    2: "Kinda",
+    3: "Quite",
+    4: "Very",
+    5: "Totally"
+  };
+  Object.keys(ConfidenceLevel).map((key, index) => {
```
[the remark this answers is not in the document]
Comment (drewjhart) [reply host/src/components/ConfidenceResponses/ConfidenceResponseGraph.jsx:75]: fixed

File: host/src/containers/GameSessionContainer.tsx (hunk 1 above)
Comment (drewjhart) [reply host/src/containers/GameSessionContainer.tsx:237]: Added the `isNullOrDefault` check above and removed the `?`

Comment (drewjhart) [reply host/src/containers/GameSessionContainer.tsx:237]: way cleaner, fixed

File: host/src/containers/GameSessionContainer.tsx (hunk 3 above)
Comment (drewjhart) [reply host/src/containers/GameSessionContainer.tsx:260]: yup fixed

**Pull request merged**